govcd/nsxt_distributed_firewall.go

```diff
@@ -99,3 +109,329 @@ func (firewall *DistributedFirewall) DeleteAllRules() error {
 
 	return firewall.VdcGroup.DeleteAllDistributedFirewallRules()
 }
+
+// GetDistributedFirewallRuleById retrieves single Distributed Firewall Rule by ID
+func (vdcGroup *VdcGroup) GetDistributedFirewallRuleById(id string) (*DistributedFirewallRule, error) {
+	if id == "" {
+		return nil, fmt.Errorf("id must be specified")
+	}
+
+	client := vdcGroup.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointVdcGroupsDfwRules
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// "default" policy is hardcoded because there is no other policy supported
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, vdcGroup.VdcGroup.Id, types.DistributedFirewallPolicyDefault), "/", id)
+	if err != nil {
+		return nil, err
+	}
+
+	returnObject := &DistributedFirewallRule{
+		Rule:     &types.DistributedFirewallRule{},
+		client:   client,
+		VdcGroup: vdcGroup,
+	}
+
+	err = client.OpenApiGetItem(apiVersion, urlRef, nil, returnObject.Rule, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving Distributed Firewall rule: %s", err)
+	}
+
+	return returnObject, nil
+}
+
+// GetDistributedFirewallRuleByName retrieves single firewall rule by name
+func (vdcGroup *VdcGroup) GetDistributedFirewallRuleByName(name string) (*DistributedFirewallRule, error) {
+	if name == "" {
+		return nil, fmt.Errorf("name must be specified")
+	}
+
+	dfw, err := vdcGroup.GetDistributedFirewall()
+	if err != nil {
+		return nil, fmt.Errorf("error returning distributed firewall rules: %s", err)
+	}
+
+	var filteredByName []*types.DistributedFirewallRule
+	for _, rule := range dfw.DistributedFirewallRuleContainer.Values {
+		if rule.Name == name {
+			filteredByName = append(filteredByName, rule)
+		}
+	}
+
+	oneByName, err := oneOrError("name", name, filteredByName)
+	if err != nil {
+		return nil, err
+	}
+
+	return vdcGroup.GetDistributedFirewallRuleById(oneByName.ID)
+}
+
+// CreateDistributedFirewallRule is a wrapper around "vdcGroups/%s/dfwPolicies/%s/rules" endpoint
+// which handles all distributed firewall (dfw) rules at once. While there is no real endpoint to
+// create single firewall rule, it is a requirements for some cases (e.g. using in Terraform)
+// The code works by doing the following steps:
+//
+// 1. Getting all Distributed Firewall Rules and storing them in private intermediate
+// type`distributedFirewallRulesRaw` which holds a []json.RawMessage (text) instead of exact types.
```

Comment:
An awesome idea!

Reply:
Thanks. To be fair this is not a new mechanism - we use the same mechanism for the core OpenAPI support functionality `OpenApiGetAllItems` https://github.com/vmware/go-vcloud-director/blob/main/govcd/openapi.go#L98-L112